cnd/src/db/mod.rs

```diff
@@ -0,0 +1,173 @@
+mod models;
+mod schema;
+
+pub use crate::db::models::Metadata;
+
+use crate::{
+    db::models::NewMetadata,
+    swap_protocols::{
+        metadata_store::{self, AssetKind, Error, LedgerKind, MetadataStore, Role},
+        SwapId,
+    },
+};
+use diesel::{self, prelude::*, sqlite::SqliteConnection};
+use libp2p::PeerId;
+use migrations_internals;
+use std::{
+    convert::TryFrom,
+    fs::File,
+    io::stdout,
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+
+/// This module provides an Sqlite backed persistent MatadataStore.
+
+#[derive(Debug)]
+pub struct SqliteMetadataStore {
+    db: PathBuf,
+}
+
+impl SqliteMetadataStore {
+    /// Return a handle that can be used to access the database.
+    ///
+    /// When this returns an Sqlite database exists at 'db' and connection to
+    /// the database has been verified.
+    pub fn new(db: Option<PathBuf>) -> Result<SqliteMetadataStore, Error> {
+        let db = db
+            .or(default_db_path())
+            .ok_or_else(|| Error::Path("Failed to get database path".to_string()))?;
+
+        create_database_if_needed(&db)?;
+        log::info!("Sqlite database file: {}", db.display());
+
+        let migrations = default_migrations_dir()
+            .ok_or_else(|| Error::Path("Failed to get migrations directory".to_string()))?;
+
+        create_migrations_directory_if_needed(&migrations)?;
+        run_migrations(&db, &migrations)?;
+
+        Ok(SqliteMetadataStore {
+            db: db.clone().to_path_buf(),
+        })
+    }
+}
+
+impl MetadataStore for SqliteMetadataStore {
+    fn insert<M: Into<metadata_store::Metadata>>(
+        &self,
+        metadata: M,
+    ) -> Result<(), metadata_store::Error> {
+        use schema::metadatas;
+
+        let md = Metadata::new(metadata.into());
+        let new = NewMetadata::new(&md);
+
+        let conn = establish_connection(&self.db)?;
+        diesel::insert_into(metadatas::table)
+            .values(new)
+            .execute(&conn)
+            .map(|_| ()) // Drops the '1' for successful row insert.
+            .map_err(|err| Error::Insert(err.to_string()))
+    }
+
+    fn get(&self, key: SwapId) -> Result<Option<metadata_store::Metadata>, Error> {
+        use self::schema::metadatas::dsl::*;
+
+        let key = key.to_string();
+        let conn = establish_connection(&self.db)?;
+        let records = metadatas
+            .filter(swap_id.eq(key))
+            .load::<Metadata>(&conn)
+            .map_err(|err| Error::Load(err.to_string()))?;
```

Comment:
I was wondering if there is no nicer way and I found this: 

```
metadatas
.filter(swap_id.eq(key))
.first(conn)
.optional()
```

wouldn't that work for you?

Reply:
Cool, I'll give this a shot.  I copied this from the docs.

Reply:
I got this
```
        metadatas
            .filter(swap_id.eq(key))
            .first(&conn)
            .optional()
            .map_err(|err| Error::Load(err.to_string()))?
            .map(move |m: Metadata| metadata_store::Metadata::try_from(m.clone()))
            .transpose()
```

Reply:
not sure what `transpose` does ... 
but if it works, cool :)

Reply:
`transpose` converts between `Result<Option>` and `Option<Result>` IIRC: https://doc.rust-lang.org/std/result/enum.Result.html#method.transpose